### PR TITLE
Speedup expand

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by certik** [comment]
Thanks, I wanted to extract all the things from Basic and put it outside of the core into the visitor classes or single dispatch. How can it be faster though? I though the visitor pattern has 2 virtual function calls instead of one. What are the timings on your computer?

File: symengine/expand.cpp [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+            Add::dict_add_term(d_, coef2, t);
+        }
+    }
+};
+
+//! Expands `self`
+RCP<const Basic> expand(const RCP<const Basic> &self) {
+    ExpandVisitor v;
+    v.apply(self);
+    return Add::from_dict(v.coeff, std::move(v.d_));
```
Comment (certik) [inline symengine/expand.cpp]: The last line should probably be a method in the `ExpandVisitor` and then `d_` and other variables should be private to the `ExpandVisitor` class.

**Comment by certik** [comment]
I wonder if it is faster due to using a visitor pattern instead of:
```c++
RCP<const Basic> expand(const RCP<const Basic> &self)
{
    if (is_a<Symbol>(*self)) return self;
    if (is_a_Number(*self)) return self;
    if (is_a<Add>(*self)) return add_expand(rcp_static_cast<const Add>(self));
    if (is_a<Mul>(*self)) return mul_expand(rcp_static_cast<const Mul>(self));
    if (is_a<Pow>(*self)) return pow_expand(rcp_static_cast<const Pow>(self));
    return self;
} 
```
Which indeed seems slow.

I like this PR, I think that's the way to go.

File: symengine/expand.cpp [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+    RCP<const Number> coeff = zero;
+    RCP<const Number> multiply = one;
+
+    ExpandVisitor() : BaseVisitor<ExpandVisitor>(this) {
+
+    }
+
+    void apply(const RCP<const Basic> &b) {
+        b->accept(*this);
+    }
```
Comment (certik) [inline symengine/expand.cpp]: This method is not needed, is it? I would remove it.

**Comment by isuruf** [comment]
There are 2 reasons for the speedup.

1. Due to visitor pattern instead of checking type code. 
As in #384, we see a speedup due to this.

2. We are not creating `Add` objects every time we do an expand. They are created only at the last moment and that is the main reason to the speedup seen in `expand6`

File: symengine/expand.cpp (hunk 2 above)
Comment (isuruf) [inline symengine/expand.cpp]: It was used at https://github.com/symengine/symengine/pull/668/files#diff-3ca70833b804e0d56ab69290843badb8R48,.
There were more uses in previous work, but now that there's only one place I can use the method without RCP and delete this method.

Comment (certik) [inline symengine/expand.cpp]: I would remove it.

**Comment by certik** [comment]
I see. Where were we creating `Add` objects before? I thought we constructed the dictionary and then passed it to `Add`.

**Comment by isuruf** [comment]
For eg. `(x+y)**2*y-(x+y)**3` here we first expanded the first term in the add, then expanded the second term and then added those expanded terms together.

New strategy is to expand the first term and insert it to the dictionary, then expand the second term and insert to the same dictionary and finally create an Add.

**Comment by certik** [comment]
Ah, I didn't realize that. Yes, that's much better.

**Comment by isuruf** [comment]
This is ready for review now

**Comment by certik** [comment]
This is +1 from me. Let me rerun tests, to make sure they still all pass.

**Comment by certik** [comment]
Looks like tests still pass, merging.

**Pull request merged**